compiler/pipes/check-abstract-function-defaults.cpp

```diff
@@ -0,0 +1,49 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2020 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#include "compiler/pipes/check-abstract-function-defaults.h"
+
+#include "common/containers/final_action.h"
+#include "common/termformat/termformat.h"
+
+#include "compiler/data/class-data.h"
+#include "compiler/data/function-data.h"
+#include "compiler/pipes/calc-real-defines-values.h"
+
+void CheckAbstractFunctionDefaults::execute(FunctionPtr interface_function, DataStream<FunctionPtr> &os) {
+  stage::set_function(interface_function);
+  auto forward_function_next = vk::finally([&] { os << interface_function; });
+
+  if (!interface_function->modifiers.is_abstract()) {
+    return;
+  }
+
+  auto get_default = [](VertexPtr v) {
+    auto param = v.try_as<op_func_param>();
+    kphp_assert(param);
+    if (param->has_default_value()) {
+      return param->default_value();
+    }
+    return VertexPtr{};
+  };
+
+  auto interface_params = interface_function->get_params();
+  for (auto derived_class : interface_function->class_id->get_all_inheritors()) {
+    auto *derived_method = derived_class->members.get_instance_method(interface_function->local_name());
+    if (!derived_method || derived_method->function == interface_function) {
+      continue;
+    }
+    auto derived_params = derived_method->function->get_params();
+
+    for (auto arg_id = interface_function->get_min_argn(); arg_id < interface_params.size(); ++arg_id) {
+      auto derived_default = get_default(derived_params[arg_id]);
+      auto interface_default = get_default(interface_params[arg_id]);
+      kphp_error(deep_equal(derived_default, interface_default),
```

Comment:
Please, consider the following example:
```php
<?php

interface II {
    function foo($x = 1234);
}

class AA implements II {
    function foo($x = 56789) { var_dump($x); }
}

class BB implements II {
    function foo($x = 0) { var_dump($x); }
}

function foo(II $ii) { $ii->foo(); }

function demo() {
    foo(new AA);
    foo(new BB);
}

demo();
```

Old code (before this patch) forbids such kind of default value redefining, but new one allows and generates inconsistent with PHP result
PHP out:
```
int(56789)
int(0)
```
new KPHP out:
```
int(1234)
int(1234)
```
:(

Reply:
It's a bug, thx, fixed it